packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -70,34 +80,115 @@ impl<K> Default for DependencyMap<K> {
-    ///   will be returned in order to keep resolving it
-    /// - Otherwise, `None` will be returned as no further resolution is needed
-    pub fn insert(
+    /// provided `new_resolve_depth` and `new_interval`. If the dependency was already resolved it
+    /// is checked whether the new resolve depth and interval are more general than the existing
+    /// resolve depth and interval. If they are, the existing resolve depth and interval are updated
+    /// to cover the new resolve depth and interval.
+    ///
+    /// If the traversed entry has to be resolved further, [`DependencyStatus::Unresolved`] is
+    /// returned with the new resolve depth and interval that the traversal should continue with.
+    pub fn update(
         &mut self,
         identifier: &K,
-        resolved_depth: GraphResolveDepths,
+        new_resolve_depth: GraphResolveDepths,
+        new_interval: TimeInterval<ProjectedTime>,
     ) -> DependencyStatus {
         match self.resolved.raw_entry_mut().from_key(identifier) {
             RawEntryMut::Vacant(entry) => {
-                entry.insert(identifier.clone(), resolved_depth);
-                DependencyStatus::Unresolved
+                entry.insert(
+                    identifier.clone(),
+                    (new_resolve_depth, new_interval.clone()),
+                );
+                DependencyStatus::Unresolved(new_resolve_depth, new_interval)
             }
             RawEntryMut::Occupied(entry) => {
-                if entry.into_mut().update(resolved_depth) {
-                    DependencyStatus::Unresolved
-                } else {
+                let (current_depths, current_interval) = entry.into_mut();
+                let old_interval = current_interval.clone();
+                *current_interval = current_interval.clone().merge(new_interval.clone());
+
+                if current_depths.update(new_resolve_depth) {
+                    // It does not matter if the interval is updated or not, as it has to be
+                    // resolved for the full interval anyway.
+                    DependencyStatus::Unresolved(*current_depths, current_interval.clone())
+                } else if old_interval.contains_interval(&new_interval) {
+                    tracing::info!("{old_interval:?} contains {new_interval:?}");
+                    // The dependency is already resolved for the required interval
+                    // old: [-----)
+                    // new:  [---)
                     DependencyStatus::Resolved
+                } else if new_interval.contains_interval(&old_interval)
+                    || new_interval.is_adjacent_to(&old_interval)
+                {
+                    // The dependency is already resolved, but not for the required interval or the
+                    // intervals are adjacent so we only have to resolve the new interval
+                    // old:  [---)   |  [---)
+                    // new: [-----)  |      [---)
```

Comment:
This is another location where I think we should highlight that this isn't _necessarily_ optimum, and in the first case we'd ideally just resolve the symmetric difference.

Reply:
Improved the comment: 14e2eb59923d783be881a75cc221d065e038cf6f

Reply:
This one's great!